mint-client/src/clients/user.rs

```diff
@@ -420,3 +452,190 @@ impl From<LnClientError> for ClientError {
+    where
+        E: serde::de::Error,
+    {
+        match v.parse::<lightning_invoice::Invoice>() {
+            Ok(bolt11) => Ok(WrappedInvoice(bolt11)),
+            Err(e) => Err(serde::de::Error::custom(e.to_string())),
+        }
+    }
+}
+
+impl<'de> serde::de::Visitor<'de> for PegOutReqVisitor {
+    type Value = PegOutReq;
+
+    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+        formatter.write_str("Could not deserialize pegout request")
+    }
+    fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: serde::de::MapAccess<'de>,
+    {
+        let mut addr: Option<&str> = None;
+        let mut amt: Option<u64> = None;
+
+        while let Some(key) = access.next_key::<String>()? {
+            match key.as_str() {
+                "address" => addr = Some(access.next_value()?),
+                "amount" => amt = Some(access.next_value()?),
+                _ => (),
+            }
+        }
+        Ok(PegOutReq {
+            address: bitcoin::Address::from_str(addr.unwrap()).unwrap(), //unwraps are not ok here I think but I will come back to Errors in another commit anyway
+            amount: bitcoin::Amount::from_sat(amt.unwrap()),
+        })
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for WrappedInvoice {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(InvoiceVisitor {})
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for PegOutReq {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_map(PegOutReqVisitor {})
+    }
+}
+///Holds either a Success or an Error variant and will be send as a response to the client (note: this means all [`ClientError`] will be send as [`tide::Result`] Ok
+/// only other errors will cause the server to respond with [`tide::Result`] Err)
+pub enum Event {
+    Success(APIResponse),
+    Error(APIResponse), //TODO: define Error APIResponse variant, implement a build method and use it in the endpoints
+}
+/// Holds all possible Responses of the RPC-CLient can also be used to parse responses (for client-cli)
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub enum APIResponse {
+    ///The clients holdings : The quantity of coins for each tier. For total holdings sum(Infoi.quantity * Infoi.tier) with i = 0 - n
+    /// Also contains the [`APIResponse::Pending`] variant.
+    Info {
+        coins: Vec<CoinsByTier>,
+        pending: Box<APIResponse>,
+    },
+    /// Active issuances : Not yet (bey the federation) signed BUT accepted coins
+    Pending {
+        //TODO: Also return Vec<TransactionId> (?)
+        transactions: usize,
+        acc_qty_coins: usize,
+        acc_val_amount: Amount,
+    },
+    ///Holds a new address for the client to use for peg-in
+    PegInAddress { pegin_address: String },
```

Comment:
Can this use `bitcoin::Address`? Should be possible if you activate serde for rust-bitcoin.